espnet2/train/preprocessor.py

```diff
@@ -489,3 +493,203 @@ def _text_process(
                 data[text_name] = np.array(text_ints, dtype=np.int64)
         assert check_return_type(data)
         return data
+
+
+class EnhPreprocessor(CommonPreprocessor):
+    """Preprocessor for Speech Enhancement (Enh) task."""
+
+    def __init__(
+        self,
+        train: bool,
+        rir_scp: str = None,
+        rir_apply_prob: float = 1.0,
+        noise_scp: str = None,
+        noise_apply_prob: float = 1.0,
+        noise_db_range: str = "3_10",
+        speech_volume_normalize: float = None,
+        speech_name: str = "speech_mix",
+        speech_ref_name_prefix: str = "speech_ref",
+        noise_ref_name_prefix: str = "noise_ref",
+        dereverb_ref_name_prefix: str = "dereverb_ref",
+        use_reverberant_ref: bool = False,
+        num_spk: int = 1,
+        num_noise_type: int = 1,
+        sample_rate: int = 8000,
+        force_single_channel: bool = False,
+    ):
+        super().__init__(
+            train=train,
+            token_type=None,
+            token_list=None,
+            bpemodel=None,
+            text_cleaner=None,
+            g2p_type=None,
+            unk_symbol="<unk>",
+            space_symbol="<space>",
+            non_linguistic_symbols=None,
+            delimiter=None,
+            rir_scp=rir_scp,
+            rir_apply_prob=rir_apply_prob,
+            noise_scp=noise_scp,
+            noise_apply_prob=noise_apply_prob,
+            noise_db_range=noise_db_range,
+            speech_volume_normalize=speech_volume_normalize,
+            speech_name=speech_name,
+        )
+        self.speech_ref_name_prefix = speech_ref_name_prefix
+        self.noise_ref_name_prefix = noise_ref_name_prefix
+        self.dereverb_ref_name_prefix = dereverb_ref_name_prefix
+        self.use_reverberant_ref = use_reverberant_ref
+        self.num_spk = num_spk
+        self.num_noise_type = num_noise_type
+        self.sample_rate = sample_rate
+        self.force_single_channel = force_single_channel
+
+    def _ensure_2d(self, signal):
+        if isinstance(signal, tuple):
+            return tuple(self._ensure_2d(sig) for sig in signal)
+        elif isinstance(signal, list):
+            return [self._ensure_2d(sig) for sig in signal]
+        else:
+            # (Nmic, Time)
+            return signal[None, :] if signal.ndim == 1 else signal.T
+
+    def _get_early_signal(self, speech, rir):
+        predelay = 50  # milliseconds
```

Comment:
Correct way to deal with misalignment when you don't know the anechoic RIR ! I like this.

Reply:
Oh yes. I forgot to do it! Thanks for reminding me of that.